[videoplayer] allow custom headers to be sent to ffmpeg & inputstream class property

phunkyfish opened the pull request:
## Description
<!--- Provide a general summary of your change in the Pull Request title above -->
<!--- Describe your change in detail here -->

Currently all unknown headers are not allowed to be passed to ffmpeg. This PR includes a blacklist to stop certain headers ever being sent to ffmpeg ad a whitelist of other ffmpeg headers. This allows all other custom headers to be passed.

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here -->

## How Has This Been Tested?
<!--- Please describe in detail how you tested your change -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc -->

On OSX, verifying that headers get/don't get passed from teh log.


## Screenshots (if appropriate):

## Types of change
<!--- What type of change does your code introduce? Put an `x` in all the boxes that apply like this: [X] -->
- [ ] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [X] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

## Checklist:
<!--- Go over all the following points, and put an `X` in all the boxes that apply like this: [X] -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [X] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [X] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

Rechi commented on xbmc/cores/VideoPlayer/DVDDemuxers/DVDDemuxFFmpeg.cpp:
```diff
@@ -782,12 +782,30 @@ AVDictionary* CDVDDemuxFFmpeg::GetFFMpegOptionsFromInput()
+               name == "reconnect_delay_max" || name == "icy" || name == "icy_metadata_headers" ||
+               name == "icy_metadata_packet")
+      {
+        CLog::Log(LOGDEBUG, "CDVDDemuxFFmpeg::GetFFMpegOptionsFromInput() adding custom header option '%s: %s'", it->first.c_str(), value.c_str());
```
> The `ColumnLimit` in `.clang-format` is set to `100` which defines line length [...]

https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md#3-formatting

Rechi commented on xbmc/cores/VideoPlayer/DVDDemuxers/DVDDemuxFFmpeg.cpp:
```diff
@@ -782,12 +782,30 @@ AVDictionary* CDVDDemuxFFmpeg::GetFFMpegOptionsFromInput()
+               name == "offset" || name == "end_offset" || name == "method" || name == "listen" ||
+               name == "send_expect_100")
+      {
+        CLog::Log(LOGDEBUG, "CDVDDemuxFFmpeg::GetFFMpegOptionsFromInput() ignoring header option '%s'", it->first.c_str());
```
> The `ColumnLimit` in `.clang-format` is set to `100` which defines line length [...]

https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md#3-formatting

Rechi commented on xbmc/cores/VideoPlayer/DVDDemuxers/DVDDemuxFFmpeg.cpp:
```diff
@@ -782,12 +782,30 @@ AVDictionary* CDVDDemuxFFmpeg::GetFFMpegOptionsFromInput()
       else
       {
-        CLog::Log(LOGDEBUG, "CDVDDemuxFFmpeg::GetFFMpegOptionsFromInput() ignoring header option '%s: %s'", it->first.c_str(), value.c_str());
+        CLog::Log(LOGDEBUG, "CDVDDemuxFFmpeg::GetFFMpegOptionsFromInput() adding user custom header option '%s: ***********'", it->first.c_str());
```
> The `ColumnLimit` in `.clang-format` is set to `100` which defines line length [...]

https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md#3-formatting

phunkyfish commented on xbmc/cores/VideoPlayer/DVDDemuxers/DVDDemuxFFmpeg.cpp:
```diff
@@ -782,12 +782,30 @@ AVDictionary* CDVDDemuxFFmpeg::GetFFMpegOptionsFromInput()
+      {
+        CLog::Log(LOGDEBUG, "CDVDDemuxFFmpeg::GetFFMpegOptionsFromInput() ignoring header option '%s'", it->first.c_str());
+      }
+      // Any other headers would be user defined and should be safe to pass
```
The main example I found was http_proxy and that makes perfect sense. I was trying to give a method to send headers that were intended to be sent to the outside.

Curl uses a `!` notation to denote a header that should be sent. I guess that is better than just letting anything through.

phunkyfish replied:
If we use `!` it's less obvious but at least consistent with Curl:

https://github.com/xbmc/xbmc/blob/master/xbmc/filesystem/CurlFile.cpp#L835

phunkyfish replied:
Easy to swap `!` for any other prefix.

FernetMenta replied:
the "!" in the linked code only prevents from logging and does not address the original problem. blindly sending everything as a header is wrong and @peak3d re-introduced the issue again.

FernetMenta replied:
"custom headers" are not wrong. the problem is that at the time the headers are set you don't know if an options was supposed to be a header.

phunkyfish replied:
But we’re ok with the approach for ffmpeg?

FernetMenta replied:
Depends how much you care about security. You can flag options explicitly as headers to be sent or you can just send everything by assumption. imo this is a very simple decision.

phunkyfish replied:
I think we need to balance security with the openness of the platform. I think this PR still prevents users from unknowingly sending data to the internet which they did not intend and also allows them to be explicit in what they do send.

FernetMenta replied:
> I think we need to balance security with the openness of the platform

I don't get this. If you prefix the options as I suggested the platform is as open as you desire but more secure.

phunkyfish replied:
Sorry, my mistake. I never pushed the update ;)

FernetMenta replied:
ahh, makes sense now :)

Rechi commented on xbmc/cores/VideoPlayer/DVDDemuxers/DVDDemuxFFmpeg.cpp:
```diff
@@ -782,12 +782,35 @@ AVDictionary* CDVDDemuxFFmpeg::GetFFMpegOptionsFromInput()
       else
       {
-        CLog::Log(LOGDEBUG, "CDVDDemuxFFmpeg::GetFFMpegOptionsFromInput() ignoring header option '%s: %s'", it->first.c_str(), value.c_str());
+        CLog::Log(LOGDEBUG, "CDVDDemuxFFmpeg::GetFFMpegOptionsFromInput() ignoring header option '%s'", it->first.c_str());
```
> The `ColumnLimit` in `.clang-format` is set to `100` which defines line length [...]

https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md#3-formatting

phunkyfish commented:
@FernetMenta I added a commit here which would allow channels played from PVR addons such as iptvsimple to also pass ffmpeg headers (similar to using strm files).

Is it reasonable to assume that DVDSTREAM_TYPE_FFMPEG will function as well as DVDSTREAM_TYPE_FILE due to this commit? Happy to remove it if there are more serious consequences to the change.

a1rwulf commented on xbmc/cores/VideoPlayer/DVDInputStreams/DVDFactoryInputStream.cpp:
```diff
@@ -150,6 +150,9 @@ std::shared_ptr<CDVDInputStream> CDVDFactoryInputStream::CreateInputStream(IVide
       }
     }
 
+    if (URIUtils::IsProtocol(finalFileitem.GetPath(), "pvr"))
```
Mhm not sure if `GetPath()` is correct here.
You probably need to check the `DynPath`.
@ksooo probably knows?

phunkyfish replied:
The intention was to pick up streams that were not already detected as DVDSTREAM_TYPE_PVRMANAGER. From testing they always appeared to preset satisfying the above condition.

ksooo replied:
path == item identity (must not change during item instance' lifetime)
dynpath == "playback url" (can change at any time, thus the name)

most of the time path == dynpath

Examples: 
* pvr.hts (own demuxer) recordings/channels. path = dynpath = "pvr://..."
* pvr.iptvsimple (no own demuxer): recordings/channels. path = "püvr://...", dynpath = "http(s)://..."

FernetMenta commented:
@phunkyfish your code targets pvr items that have set a dynPath. Since dynPath is not limited to internet streams only,  this change is wrong.

phunkyfish commented:
@FernetMenta, can you explain a little further for my understanding? 

I'm trying to target pvr items that are internet streams, i.e. could require the ability to pass http headers. Is there another class of items with dynPath that I need to take care of?

FernetMenta commented:
dynPath can be anything but you don't look at it. your current code assumes that dynPath for pvr items is always an internet stream url that can be handled by ffmpeg. those items with a dynPath that can be handled by ffmpeg should already be considered by the factory.
hard to say more without seeing an example url.

phunkyfish commented:
Ok, that makes sense. However the factory is not handling any of them currently. For instance `http(s)` would not be handled as it is.

Ideally there would be a whitelist of `dynPath` protocols that would be supported by ffmpeg. This should be easy enough to add. Thanks for the help ;)

FernetMenta commented:
> For instance http(s) would not be handled as it is.

http(s) is not supposed to be handled by ffmpeg. passing those to inputstream  ffmpeg would break a couple of functions like shoutcast redirect or file caching.

> Ideally there would be a whitelist of dynPath protocols that would be supported by ffmpeg

already done by the factory

phunkyfish commented:
Ok, but then what would be the best way to pass http headers if not using inputstream ffmpeg?

FernetMenta commented:
why don't those headers find their why to CurlFile?

phunkyfish commented:
They do find there way to CurlFile. But they don’t find their way beyond that to FFmpeg such as for HLS and video fragments.

FernetMenta commented:
If CurlFile handles the http protocol and inputstream, ffmpeg does only demuxing. http headers are meaningless for a demuxer.
HLS is identified by the m3u8 extension and passed to inputstream ffmpeg.

you seem to have an odd or new case but you don't give much information on that.

phunkyfish commented:
Sorry, never posted the reply. The reason is because the extension needs to be .m3u8 for it to choose inputstream fffmpeg. In my use case I'm serving the file from an API and it doesn't have this extension.

FernetMenta commented:
You are facing the same issue like others before. I don't like the "solutions" for this in Kodi but see yourself:

For shoutcast streams, which are served over http(s) too, are new protocol specifier was introduced:
shout://blabla
This is just as little a protocol as pvr:// is and has the same weakness: there are cases where the url lacks this protocol specifier. 
If CurlFile i.e. opens a shoutcast URL, it fires this ugly CRedirectException

Nowadays players get the entire fileitem insead just the url. IMO a much better solution is to set a property on the fileitem that can be examined by the factory. That's btw already done to select particular inputstream addons.

phunkyfish commented:
Perfect, I can use the latter to set a property via the PVR addon easily. That's makes the solution nice and clean.

ksooo commented:
> Perfect, I can use the latter to set a property via the PVR addon easily.

 This proerty should then get part of the PVR addon API and needs to be documented properly:

https://github.com/xbmc/xbmc/blob/master/xbmc/addons/kodi-addon-dev-kit/include/kodi/xbmc_pvr_types.h#L71 ff

ksooo commented:
Maybe even the existing "mimetype" property could be used here?

https://github.com/xbmc/xbmc/blob/master/xbmc/addons/kodi-addon-dev-kit/include/kodi/xbmc_pvr_types.h#L70

phunkyfish commented:
Which would you prefer? I added it as a separate property in latest commit. I made it general and not specific to m3u8, just setting a preference for inputstream ffmpeg.

FernetMenta commented:
As I was active I planned to split the streaming API from the rest of the PVR API. Addons would request an inputstream class with this method. There is already a property for this PVR_STREAM_PROPERTY_INPUTSTREAMADDON. I don't see any advantage in limiting this to addons. Why don't you drop "addon" and name it "inputstreamclass"

phunkyfish commented:
You may not have an inputstream addon installed so having it separate serves as a fallback.

ksooo commented:
I was thinking almost the same. Isn't the internal demuxer/player (ffmpeg) just another "input steam addon". So, my suggestion is to use PVR_STREAM_PROPERTY_INPUTSTREAMADDON (without renaming it for now) and to define a unique name for the kodi-internal "ffmpeg addon".

phunkyfish commented:
Cool, ya, let’s do that.

FernetMenta commented:
> Isn't the internal demuxer/decoder (ffmpeg) just another "input stream addon".

it is no addon and abusing something called "addon" for internal classes is wrong and misleading. doing this cases deterioaration of the pvr api AND vp. 

```
    if fileitem.GetProperty("inputstreamaaddon").asString() == "ffmpeg)
      return std::shared_ptr<CDVDInputStreamFFmpeg>(new CDVDInputStreamFFmpeg(finalFileitem));
```

I don't see any advantage in such a hack.

ksooo commented:
> I don't see any advantage in such a hack.

 And if we rename the property from PVR_STREAM_PROPERTY_INPUTSTREAMADDON  to PVR_STREAM_PROPERTY_INPUTSTREAMCLASS (your suggestion) it is no longer a hack? Nothing else would change except the name of the property, hmmm.

FernetMenta commented:
You should keep in mind that requesting an inputstream class in NOT a pvr thingy. Setting this property is for all types of addons. Renaming the property would break addons like amazon.video. The best solution is adding a new property "inputstreamclass". Keep both, then depreciate the old one after a while.

I am referring to code in VP. Renaming the PVR define is easy, because it is a define. AFAIK other addons do not use a define.

ksooo commented:
Makes sense.

phunkyfish commented:
Ok, so like the latest commit?

Rechi commented on xbmc/cores/VideoPlayer/DVDInputStreams/DVDFactoryInputStream.cpp:
```diff
@@ -153,6 +153,9 @@ std::shared_ptr<CDVDInputStream> CDVDFactoryInputStream::CreateInputStream(IVide
     if (finalFileitem.IsType(".m3u8"))
       return std::shared_ptr<CDVDInputStreamFFmpeg>(new CDVDInputStreamFFmpeg(finalFileitem));
 
+    if (finalFileitem.IsPVR() && fileitem.GetProperty("inputstreamclass").asString() == "inputstream.ffmpeg")
```
> The `ColumnLimit` in `.clang-format` is set to `100` which defines line length [...]

https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md#3-formatting

FernetMenta replied:
this is not a pvr thingy -> test for pvr is wrong.
explicit requests for classes need to be done right after addons are checked -> move this up

FernetMenta replied:
"inputstream.ffmpeg" this should be documented in some api file. how should an addon dev know how internal classes are named?

FernetMenta replied:
check out cores/VideoPlayer/Interface/Addon/

phunkyfish replied:
Sorry, meant to remove PVR check. If I move it up with misses the Curl check to set the DynPath.

FernetMenta replied:
> If I move it up it misses the Curl check to set the DynPath.

fileitems are passed to inputstream, that includes dynPath. that also works for other inputstream classes like addons.

phunkyfish replied:
Ok, works fine for internet based m3u8 files. If I copy one local I get:
`ffmpeg[0x7fd13dc20a40X]: [https] Protocol 'https' not on whitelist 'file,crypto'!`

phunkyfish replied:
It feels wrong to limit `inputstreamclass` to internet streams only.

phunkyfish replied:
Ok, added  `av_dict_set(&options, "protocol_whitelist", "file,http,https,tcp,tls,crypto", 0);` as the whitelist in `GetFFMpegOptionsFromInput` and it works with local m3u8 files. 

However I don't know if this is acceptable or not.

phunkyfish commented:
Ok, latest allows local and remote files to use inputstream ffmpeg. I added a header to define all the inputstream properties. 

Should I also define the likes of `program` in this header? Any others that are undefined elsewhere?

FernetMenta commented on xbmc/cores/VideoPlayer/DVDDemuxers/DVDDemuxFFmpeg.cpp:
```diff
@@ -725,6 +725,8 @@ AVDictionary* CDVDDemuxFFmpeg::GetFFMpegOptionsFromInput()
   CURL url = m_pInput->GetURL();
   AVDictionary* options = nullptr;
 
+  av_dict_set(&options, "protocol_whitelist", "file,http,https,tcp,tls,crypto", 0);
```
don't you think there is a reason why ffmpeg implemented this option? blindly bypassing it for all cases does not look right.

phunkyfish replied:
You should be able to have any protocols in a local m3u8 file. So I guess it should be all if not an internet stream and left as default otherwise.

FernetMenta commented on xbmc/cores/VideoPlayer/Interface/Addon/InputStream.h:
```diff
@@ -0,0 +1,15 @@
+
+#pragma once
+
+#define STREAM_PROPERTY_STREAMURL "streamurl" /*!< @brief the URL of the stream that should be played. */
```
again, this is not a pvr thingy. streamurl is only used by pvr and meaningless for all other addons

phunkyfish replied:
Ok, didn’t realise it was not used by other addons.

FernetMenta commented on xbmc/cores/VideoPlayer/DVDInputStreams/DVDFactoryInputStream.cpp:
```diff
@@ -57,6 +57,9 @@ std::shared_ptr<CDVDInputStream> CDVDFactoryInputStream::CreateInputStream(IVide
       return std::shared_ptr<CInputStreamAddon>(new CInputStreamAddon(addonInfo, pPlayer, fileitem));
   }
 
+  if (fileitem.GetProperty("inputstreamclass").asString() == "inputstream.ffmpeg")
```
why did you create a define and don't use it here?

phunkyfish replied:
I’ll take that one on the chin 😉

phunkyfish commented:
Ok, hope we are looking good now.

phunkyfish commented on xbmc/addons/kodi-addon-dev-kit/include/kodi/xbmc_pvr_types.h:
```diff
@@ -67,8 +67,10 @@ struct DemuxPacket;
+#define PVR_STREAM_PROPERTY_INPUTSTREAMCLASS  "inputstreamclass" /*!< @brief the name of the inputstream add-on that should be used by Kodi to play the stream denoted by PVR_STREAM_PROPERTY_STREAMURL. Leave blank to use Kodi's built-in playing capabilities or to allow ffmpeg to handle directly set to `inputstream.ffmpeg`. */
 #define PVR_STREAM_PROPERTY_MIMETYPE "mimetype" /*!< @brief the MIME type of the stream that should be played. */
 #define PVR_STREAM_PROPERTY_ISREALTIMESTREAM "isrealtimestream" /*!< @brief "true" to denote that the stream that should be played is a realtime stream. Any other value indicates that this is no realtime stream.*/
+#define PVR_STREAM_PROPERTY_VALUE_INPUTSTREAMFFMPEG  "inputstream.ffmpeg" /*!< @brief special value for PVR_STREAM_PROPERTY_INPUTSTREAMCLASS to use ffmpeg to directly play a stream URL. */
```
@ksooo as there will now be an InputStream.h file probably don't need to add these new lines here. Thoughts?

In fact, PVR_STREAM_PROPERTY_MIMETYPE and PVR_STREAM_PROPERTY_ISREALTIMESTREAM could also be deprecated in preference of the more general defines.

Rechi commented on xbmc/cores/VideoPlayer/DVDInputStreams/DVDFactoryInputStream.cpp:
```diff
@@ -6,6 +6,7 @@
  *  See LICENSES/README.md for more information.
  */
 
+#include "cores/VideoPlayer/Interface/Addon/InputStream.h"
```
> Included header files have to be sorted (case sensitive) alphabetically to prevent duplicates and allow better overview, with an empty line clearly separating sections.

https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md#7-headers

Rechi commented on xbmc/cores/VideoPlayer/DVDInputStreams/DVDFactoryInputStream.cpp:
```diff
@@ -57,6 +58,9 @@ std::shared_ptr<CDVDInputStream> CDVDFactoryInputStream::CreateInputStream(IVide
       return std::shared_ptr<CInputStreamAddon>(new CInputStreamAddon(addonInfo, pPlayer, fileitem));
   }
 
+  if (fileitem.GetProperty(STREAM_PROPERTY_INPUTSTREAMCLASS).asString() == STREAM_PROPERTY_VALUE_INPUTSTREAMFFMPEG)
```
> The `ColumnLimit` in `.clang-format` is set to `100` which defines line length [...]

https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md#3-formatting

FernetMenta commented on xbmc/cores/VideoPlayer/Interface/Addon/InputStream.h:
```diff
@@ -0,0 +1,14 @@
+
+#define STREAM_PROPERTY_INPUTSTREAMCLASS  "inputstreamclass" /*!< @brief the name of the inputstream add-on that should be used by Kodi to play the stream denoted by STREAM_PROPERTY_STREAMURL. Leave blank to use Kodi's built-in playing capabilities or to allow ffmpeg to handle directly set to STREAM_PROPERTY_VALUE_INPUTSTREAMFFMPEG. */
+#define STREAM_PROPERTY_MIMETYPE "mimetype" /*!< @brief the MIME type of the stream that should be played. */
+#define STREAM_PROPERTY_ISREALTIMESTREAM "isrealtimestream" /*!< @brief "true" to denote that the stream that should be played is a realtime stream. Any other value indicates that this is not a realtime stream.*/
```
https://github.com/xbmc/xbmc/blob/770c691bfeea88d561b39d2e38348ac434db3adc/xbmc/cores/VideoPlayer/DVDInputStreams/DVDInputStream.cpp#L17

phunkyfish replied:
Ah, I missed an isrealtimestream, whoops.

FernetMenta commented on xbmc/cores/VideoPlayer/Interface/Addon/InputStream.h:
```diff
@@ -0,0 +1,14 @@
+#pragma once
+
+#define STREAM_PROPERTY_INPUTSTREAMCLASS  "inputstreamclass" /*!< @brief the name of the inputstream add-on that should be used by Kodi to play the stream denoted by STREAM_PROPERTY_STREAMURL. Leave blank to use Kodi's built-in playing capabilities or to allow ffmpeg to handle directly set to STREAM_PROPERTY_VALUE_INPUTSTREAMFFMPEG. */
+#define STREAM_PROPERTY_MIMETYPE "mimetype" /*!< @brief the MIME type of the stream that should be played. */
```
where in VP do you see this used?

phunkyfish replied:
Ah, I missed an `isrealtimestream`, whoops.

phunkyfish replied:
Ah, you mean completely remove this as it's not used by VP.

phunkyfish committed:
[videoplayer] allow custom headers to be sent to ffmpeg

phunkyfish commented:
Ok, @FernetMenta I think we are (hopefully) finally good ;)

phunkyfish commented:
Spotted this: #include "addons/kodi-addon-dev-kit/include/kodi/addon-instance/Inputstream.h"

Confusing naming along with `InputStream.h` that I introduced. Changed name to `InputStreamConstants.h`.

phunkyfish committed:
[videoplayer] add inputstreamclass property for selecting inputstream addon or ffmpeg

FernetMenta commented on xbmc/cores/VideoPlayer/DVDDemuxers/DVDDemuxFFmpeg.cpp:
```diff
@@ -782,12 +786,35 @@ AVDictionary* CDVDDemuxFFmpeg::GetFFMpegOptionsFromInput()
+        headers.append(it->first).append(": ").append(value).append("\r\n");
+      }
+      // here are the headers supported by ffmpeg that we DON'T ALLOW to be passed through
+      // specifically we should never pass 'http_proxy' as it would be a security risk
```
this block doesn't make much sense in current state. if a header is not explicitly allowed, it is not forwarded.

phunkyfish replied:
The idea was to be explicit on what will not be sent to ffmpeg and log it so the user knows it’s not possible. 

But I see you point, there is no functional reason to have the block.

phunkyfish commented:
Jenkins build this please

kib commented:
Build failed likely due to new slave

kib commented:
Jenkins build this please

Rechi reviewed (dismissed):
Please follow the [code guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md).

phunkyfish commented:
@Rechi, done. It was hard to place the headers in order so added to a commit to reorder them.

I did not change the lines with the long comments as breaking them still results in lines longer than 100 (plus they look really bad).

phunkyfish committed:
[videoplayer] refactoring - fix header order

phunkyfish commented:
Managed to fix the header order. You happy now @Rechi ?

phunkyfish commented:
If there are no further objections I will merge this PR tomorrow.

Rechi commented:
The log lines still exceed the column limit of 100 characters.

phunkyfish commented:
Apologies @Rechi didn't include last fixup. Should be good now.

Rechi commented:
The include order of `[videoplayer] refactoring - fix header order` commit is wrong, so either drop that commit or fix it.

phunkyfish commented:
Can you explain how it’s wrong? I believed it to be correct as per the coding guidelines.

Rechi commented:
Look at the diff files in the CI build.

phunkyfish commented:
Thanks for that @Rechi, a handy thing to know for future reference ;)

Should be good now
